DQM/BeamMonitor/plugins/TKStatus.cc

```diff
@@ -1,50 +1,112 @@
+
+//
+// class declaration
+//
+class TKStatus : public edm::one::EDAnalyzer<> {
+public:
+  TKStatus(const edm::ParameterSet&);
 
-using namespace edm;
+protected:
+  void analyze(const edm::Event& e, const edm::EventSetup& c) override;
 
-TKStatus::TKStatus(const ParameterSet& ps) {
-  dcsTkFileName_ = ps.getParameter<ParameterSet>("BeamFitter").getUntrackedParameter<std::string>("DIPFileName");
+private:
+  void dumpTkDcsStatus(std::string const&, edm::RunNumber_t, std::array<bool, 6> const&);
+
+  std::string dcsTkFileName_;
+  edm::EDGetTokenT<DcsStatusCollection> dcsStatusToken_;
+  edm::EDGetTokenT<DCSRecord> dcsRecordToken_;
+
+  int lastlumi_ = -1;
+  // ----------member data ---------------------------
+};
+
+// ----------------------------------------------------------
+TKStatus::TKStatus(const edm::ParameterSet& ps)
+    : dcsStatusToken_(consumes<DcsStatusCollection>(edm::InputTag("scalersRawToDigi"))),
+      dcsRecordToken_(consumes<DCSRecord>(edm::InputTag("onlineMetaDataDigis"))) {
+  dcsTkFileName_ = ps.getParameter<edm::ParameterSet>("BeamFitter").getUntrackedParameter<std::string>("DIPFileName");
   {
     std::string tmpname = dcsTkFileName_;
     tmpname.insert(dcsTkFileName_.length() - 4, "_TkStatus");
     dcsTkFileName_ = std::move(tmpname);
   }
-  dcsStatus_ = consumes<DcsStatusCollection>(ps.getUntrackedParameter<std::string>("DCSStatus", "scalersRawToDigi"));
 }
 
 // ----------------------------------------------------------
-void TKStatus::analyze(const Event& iEvent, const EventSetup& iSetup) {
+void TKStatus::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
   int nthlumi = iEvent.luminosityBlock();
   if (nthlumi > lastlumi_) {  // check every LS
     lastlumi_ = nthlumi;
     // Checking TK status
-    Handle<DcsStatusCollection> dcsStatus;
-    iEvent.getByToken(dcsStatus_, dcsStatus);
+    edm::Handle<DcsStatusCollection> dcsStatus;
+    iEvent.getByToken(dcsStatusToken_, dcsStatus);
+    edm::Handle<DCSRecord> dcsRecord;
+    iEvent.getByToken(dcsRecordToken_, dcsRecord);
 
     std::array<bool, 6> dcsTk;
     for (auto& e : dcsTk) {
       e = true;
     }
 
-    for (auto const& status : *dcsStatus) {
-      if (!status.ready(DcsStatus::BPIX))
-        dcsTk[0] = false;
-      if (!status.ready(DcsStatus::FPIX))
-        dcsTk[1] = false;
-      if (!status.ready(DcsStatus::TIBTID))
-        dcsTk[2] = false;
-      if (!status.ready(DcsStatus::TOB))
-        dcsTk[3] = false;
-      if (!status.ready(DcsStatus::TECp))
-        dcsTk[4] = false;
-      if (!status.ready(DcsStatus::TECm))
-        dcsTk[5] = false;
+    // Check that the DCS information is available in some form
+    if (!dcsStatus.isValid() && !dcsRecord.isValid()) {
+      edm::LogPrint("TkStatus") << "No DCS information available!";
+      dumpTkDcsStatus(dcsTkFileName_, iEvent.run(), dcsTk);
     }
+
+    // If DCS status (from SCAL) not available (or empty), use DCS record
+    if ((*dcsStatus).empty()) {
```

Comment:
what about the MC case?
If there's a MC case...

Reply:
I improved the overall checking logic in a similar way of #37084